pkg/controllers/user/logging/clusterlogging.go

```diff
@@ -124,18 +127,49 @@ func (c *ClusterLoggingSyncer) createOrUpdateClusterConfigMap() error {
 	return utils.UpdateConfigMap(loggingconfig.ClusterConfigPath, loggingconfig.ClusterLoggingName, "cluster", c.configmaps)
 }
 
-func (c *ClusterLoggingSyncer) updateEmbeddedEndpoint() {
-	timeout := 3 * time.Minute
-	syncInterval := 10 * time.Second
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
-	go func(ctx context.Context, cancel context.CancelFunc) {
-		for range ticker.Context(ctx, syncInterval) {
-			err := utils.UpdateEmbeddedEndpoint(c.podLister, c.serviceLister, c.clusterLoggings)
-			if err != nil {
-				logrus.Debugf("updateEmbeddedEndpoint failed, %v", err)
-			} else {
-				cancel()
-			}
+func (c *ClusterLoggingLifecycle) provision(obj *v3.ClusterLogging) (newobj *v3.ClusterLogging, err error) {
+	logger, io := utils.NewClusterLoggingLogger(obj, c.clusterLoggings, v3.LoggingConditionProvisioned)
+	defer func() {
+		if err != nil {
+			logger.Infof("provision failed %s", err.Error())
+		}
+		defer io.Close()
+	}()
+	newobj = obj
+	logger.Infof("init namespace %s", loggingconfig.LoggingNamespace)
+	if err = utils.IniteNamespace(c.namespaces); err != nil {
+		return
+	}
+	logger.Infof("init configmap %s", loggingconfig.ClusterLoggingName)
+	if err = utils.InitConfigMap(c.configmaps); err != nil {
+		return
+	}
+	//embedded
+	if utils.GetClusterTarget(obj.Spec) == "embedded" {
+		logger.Infof("create embedded %s, %s", loggingconfig.EmbeddedESName, loggingconfig.EmbeddedKibanaName)
+		if err = utils.CreateOrUpdateEmbeddedTarget(c.deployments, c.serviceAccounts, c.services, c.roles, c.rolebindings, loggingconfig.LoggingNamespace, obj); err != nil {
+			return
+		}
+	} else {
+		logger.Infof("remove embedded %s, %s", loggingconfig.EmbeddedESName, loggingconfig.EmbeddedKibanaName)
+		if err = utils.RemoveEmbeddedTarget(c.deployments, c.serviceAccounts, c.services, c.roles, c.rolebindings); err != nil {
+			return
 		}
-	}(ctx, cancel)
+	}
+
+	logger.Infof("deploy flexvolume %s", loggingconfig.LogAggregatorName)
+	if err = utils.CreateLogAggregator(c.daemonsets, c.serviceAccounts, c.clusterRoleBindings, c.clusterLister, c.clusterName, loggingconfig.LoggingNamespace); err != nil {
+		return
+	}
+
+	logger.Infof("deploy fluentd %s", loggingconfig.FluentdName)
+	if err = utils.CreateFluentd(c.daemonsets, c.serviceAccounts, c.clusterRoleBindings, loggingconfig.LoggingNamespace); err != nil {
+		return
+	}
+
+	logger.Infof("check embedded components deployment")
+	if err = utils.UpdateEmbeddedEndpointWithRetry(c.podLister, c.serviceLister, c.clusterLoggings, c.nodeLister, c.clusterName, logger); err != nil {
```

Comment:
UpdateEmbeddedEndpointWithRetry  should be inside LoggingConditionUpdated.Do.

Reply:
Also, it is confusing to call UpdateEmbeddedEndpointWithRetry here. If embedded target is not chosen, why we need to call it?